File: sdk/typescript/src/creativeLifecycleClient.ts
/**
 * CreativeLifecycleClient — FREKCORE TypeScript SDK (STATE_7, 2026-09-03).
 * Mirrors sdk/python/frekcore_sdk/creative_lifecycle_client.py exactly.
 *
 * | Method       | Endpoint                                     |
 * |--------------|---------------------------------------------------|
 * | startGenesis   | POST /api/v1/creative-lifecycle/genesis              |
 * | getHistory       | GET  /api/v1/creative-lifecycle/{preId}               |
 *
 * See backend/creative_lifecycle/routes.py. WORKSHOP/METAMORPHOSE/
 * EMISSION/LEGACY are intentionally not wrapped this state — see
 * docs/architecture/FREKCORE_SDK_CONTRACT_V1.md's own scope note.
 */

import { FetchLike } from "./registryClient";
import { raiseForFrekStatus } from "./errors";

export interface StartGenesisOptions {
  concept?: string;
  lieu?: string;
  description?: string;
  sessionToken?: string;
  adminKey?: string;
}

export class CreativeLifecycleClient {
  private readonly baseUrl: string;
  private readonly fetchImpl: FetchLike;

  constructor(baseUrl: string, fetchImpl: FetchLike = fetch) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.fetchImpl = fetchImpl;
  }

  /**
   * POST /api/v1/creative-lifecycle/genesis. Requires a holder session or
   * an admin key, same as the server route.
   */
  async startGenesis(options: StartGenesisOptions = {}): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (options.sessionToken) headers["X-FREK-Session"] = options.sessionToken;
    if (options.adminKey) headers["X-Admin-Key"] = options.adminKey;
    const resp = await this.fetchImpl(`${this.baseUrl}/api/v1/creative-lifecycle/genesis`, {
      method: "POST",
      headers,
      body: JSON.stringify({
        concept: options.concept,
        lieu: options.lieu,
        description: options.description,
      }),
    });
    await raiseForFrekStatus(resp, "/api/v1/creative-lifecycle/genesis");
    return (await resp.json()) as Record<string, unknown>;
  }

  /** GET /api/v1/creative-lifecycle/{preId}. Public, no auth. */
  async getHistory(preId: string): Promise<Record<string, unknown>> {
    const path = `/api/v1/creative-lifecycle/${encodeURIComponent(preId)}`;
    const resp = await this.fetchImpl(`${this.baseUrl}${path}`, { method: "GET" });
    await raiseForFrekStatus(resp, path);
    return (await resp.json()) as Record<string, unknown>;
  }
}

File: sdk/python/frekcore_sdk/creative_lifecycle_client.py
"""FrekcoreCreativeLifecycleClient — D2's genesis-start + history-read
surface (`/api/v1/creative-lifecycle/*`).

| Method        | Endpoint                                   |
|----------------|----------------------------------------------|
| start_genesis   | POST /api/v1/creative-lifecycle/genesis         |
| get_history      | GET  /api/v1/creative-lifecycle/{pre_id}         |

See `backend/creative_lifecycle/routes.py`. WORKSHOP/METAMORPHOSE/
EMISSION/LEGACY (all multipart or reference-only writes) are
intentionally not wrapped this state — see
`FREKCORE_SDK_CONTRACT_V1.md`'s own scope note.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

import httpx

from .errors import raise_for_frek_status


class FrekcoreCreativeLifecycleClient:
    def __init__(
        self, base_url: Optional[str] = None, *, client: Optional[httpx.Client] = None
    ) -> None:
        if client is not None:
            self._client = client
            self._owns_client = False
        else:
            if not base_url:
                raise ValueError("base_url is required when no client is provided")
            self._client = httpx.Client(base_url=base_url.rstrip("/"))
            self._owns_client = True

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> "FrekcoreCreativeLifecycleClient":
        return self

    def __exit__(self, *exc_info: Any) -> None:
        self.close()

    @staticmethod
    def _auth_headers(
        session_token: Optional[str], admin_key: Optional[str]
    ) -> Dict[str, str]:
        headers: Dict[str, str] = {}
        if session_token:
            headers["X-FREK-Session"] = session_token
        if admin_key:
            headers["X-Admin-Key"] = admin_key
        return headers

    def start_genesis(
        self,
        *,
        concept: Optional[str] = None,
        lieu: Optional[str] = None,
        description: Optional[str] = None,
        session_token: Optional[str] = None,
        admin_key: Optional[str] = None,
    ) -> Dict[str, Any]:
        """POST /api/v1/creative-lifecycle/genesis. Requires a holder
        session or an admin key, same as the server route."""
        resp = self._client.post(
            "/api/v1/creative-lifecycle/genesis",
            json={"concept": concept, "lieu": lieu, "description": description},
            headers=self._auth_headers(session_token, admin_key),
        )
        raise_for_frek_status(resp)
        return resp.json()

    def get_history(self, pre_id: str) -> Dict[str, Any]:
        """GET /api/v1/creative-lifecycle/{pre_id}. Public, no auth --
        lifecycle history is public-readable, matching D1's own
        disclosure stance."""
        resp = self._client.get(f"/api/v1/creative-lifecycle/{pre_id}")
        raise_for_frek_status(resp)
        return resp.json()
